File: UnityScript/src/Scripts/OGSlider.js
#pragma strict

public class OGSlider extends OGWidget {
	public var sliderValue : float = 0;
	public var sliderLabel : OGLabel;
	public var backgroundHeight : float = 0.25;
	public var toPercent : boolean = true;
	public var suffix : String = "";
	

	////////////////
	// Rects
	////////////////
	private function GetThumbRect () : Rect {
		return new Rect ( drawRct.x + ( sliderValue * drawRct.width ) - ( drawRct.height / 2 ), drawRct.y, drawRct.height, drawRct.height );
	}

	private function GetBackgroundRect () : Rect {
		return new Rect ( drawRct.x, drawRct.y + ( drawRct.height / 2 ) - ( ( drawRct.height * backgroundHeight ) / 2 ), drawRct.width, drawRct.height * backgroundHeight );
	}


	////////////////
	// Mouse
	////////////////
	override function OnMouseDrag () {
		if ( sliderValue >= 0 && sliderValue <= 1 ) {
			sliderValue = ( Input.mousePosition.x - 10 - this.transform.position.x ) / this.transform.lossyScale.x;
		}
	

		sliderValue = Mathf.Clamp ( Mathf.Round ( sliderValue * 100 ) / 100, 0, 1 );

		if ( sliderLabel ) {
			if ( toPercent ) {
				sliderLabel.text = ( sliderValue * 100 ).ToString() + suffix;
			} else {
				sliderLabel.text = sliderValue.ToString() + suffix;
			}	
		}
	}

	override function OnMouseCancel () {
		GetRoot().ReleaseWidget();
	}
	
	override function OnMouseUp () {
		GetRoot().ReleaseWidget();
	}
	

	////////////////
	// Update
	////////////////	
	override function UpdateWidget () {
		// Mouse
		mouseRct = CombineRects ( GetThumbRect(), GetBackgroundRect() );
	}
	
	
	////////////////
	// Draw
	////////////////	
	override function DrawSkin () {
		// Background
		OGDrawHelper.DrawSprite ( GetBackgroundRect(), styles.basic.coordinates, drawDepth );
	
		// Thumb
		OGDrawHelper.DrawSlicedSprite ( GetThumbRect(), styles.thumb.coordinates, styles.thumb.border, drawDepth );
	}
}

File: OpenGUI/Assets/Plugins/OpenGUI/Scripts/OGSlider.cs
using UnityEngine;
using System.Collections;

public class OGSlider : OGWidget {
	public enum Orientation {
		Horizontal,
		Vertical
	}
		
	public Orientation orientation;
	public float sliderValue = 0;
	public OGLabel sliderLabel;
	public float backgroundHeight = 0.25f;
	public bool toPercent = true;
	public string suffix = "";
	

	////////////////
	// Rects
	////////////////
	private Rect GetThumbRect () {
		if ( orientation == Orientation.Horizontal ) {
			return new Rect ( drawRct.x + ( sliderValue * drawRct.width ) - ( drawRct.height / 2 ), drawRct.y, drawRct.height, drawRct.height );
		} else {
			return new Rect ( drawRct.x, drawRct.y + ( sliderValue * drawRct.height ) - ( drawRct.width / 2 ), drawRct.width, drawRct.width );
		}
	}

	private Rect GetBackgroundRect () {
		if ( orientation == Orientation.Horizontal ) {
			return new Rect ( drawRct.x, drawRct.y + ( drawRct.height / 2 ) - ( ( drawRct.height * backgroundHeight ) / 2 ), drawRct.width, drawRct.height * backgroundHeight );
		} else {
			return new Rect ( drawRct.x + ( drawRct.width / 2 ) - ( ( drawRct.width * backgroundHeight ) / 2 ), drawRct.y, drawRct.width * backgroundHeight, drawRct.height );
		}
	}


	////////////////
	// Mouse
	////////////////
	override public void OnMouseDrag () {
		if ( sliderValue >= 0 && sliderValue <= 1 ) {
			if ( orientation == Orientation.Horizontal ) {
				sliderValue = ( Input.mousePosition.x - drawRct.x ) / drawRct.width;
			} else {
				sliderValue = ( Input.mousePosition.y - drawRct.y ) / drawRct.height;
			}
		}
	

		sliderValue = Mathf.Clamp ( Mathf.Round ( sliderValue * 100 ) / 100, 0, 1 );
	}

	override public void OnMouseCancel () {
		root.ReleaseWidget();
	}
	
	override public void OnMouseUp () {
		root.ReleaseWidget();
	}
	

	////////////////
	// Update
	////////////////	
	override public void UpdateWidget () {
		// Persistent vars
		isSelectable = true;
		
		// Mouse
		mouseRct = CombineRects ( GetThumbRect(), GetBackgroundRect() );
		
		// Update data
		if ( sliderLabel ) {
			if ( toPercent ) {
				sliderLabel.text = ( sliderValue * 100 ).ToString() + suffix;
			} else {
				sliderLabel.text = sliderValue.ToString() + suffix;
			}	
		}
	}
	
	
	////////////////
	// Draw
	////////////////	
	override public void DrawSkin () {
		// Background
		OGDrawHelper.DrawSprite ( GetBackgroundRect(), styles.basic, drawDepth, tint, clipTo );
	
		// Thumb
		OGDrawHelper.DrawSlicedSprite ( GetThumbRect(), styles.thumb, drawDepth, tint, clipTo );
	}		
}
